### DISCUSSION TITLE
QF/J rolling logs with log4j2.

### DISCUSSION THREAD

**Discussion by hamilton-capital** [quickfix-j/quickfixj#558] [General]
Can anyone point me to an example of rolling logfiles using log4j2? 

There is not much on the web on this topic.

**Comment by chrjohn** [comment]
Isn't this more a log4j question than a QFJ one?
I think it should be this: https://logging.apache.org/log4j/2.x/manual/appenders.html#RollingFileAppender

**Reply by hamilton-capital** [reply]
Thanks for the reply. 

I can get the rolling of log files working with log4j2 on its own - that's not the issue. 

The problem is I don't know how to create the QFJ logs to use the log4j2 config. Are there any FIX settings that I can set to tell QFJ to use the log4j2 config to set up the files, format of logging etc.

**Reply by chrjohn** [reply]
Are you using a stand-alone Java application? Normally it should work when you put the log4j2 libs along with the config file (usually log4j2.xml) on the classpath of your app that uses QFJ. What does your classpath look like? log4j2 will not be there by default since it is not included in the QFJ archive.

**Reply by hamilton-capital** [reply]
So I have added the latest maven deps for log4j2 (see below) to my intellj project and have also added a log4j2.xml in the resources folder. I can use this logging in the same project that I have a QFJ instance created. I am a little confused how I wire up the QFJ logs to use log4j2.xml. Do I need to add something to the log4j2.xml and/of FIX settings file or something else?

<dependency>
    <groupId>org.apache.logging.log4j</groupId>
    <artifactId>log4j-core</artifactId>
    <version>2.19.0</version>
</dependency>
<dependency>
    <groupId>org.apache.logging.log4j</groupId>
    <artifactId>log4j-api</artifactId>
    <version>2.19.0</version>
</dependency>

**Reply by chrjohn** [reply]
Which implementation of QFJ's `LogFactory` are you using? Did you try with `SLF4JLogFactory`? That should use a log4j2 appender if it is on the classpath.

**Reply by hamilton-capital** [reply]
Correct, I am trying with SLF4JLogFactory. 

Would you happen to have an example FIX settings and log4j2.xml file where this is currently working so I can cross check what I am doing wrong?

**Reply by hamilton-capital** [reply]
Also, does it matter what version of log4j2 I am using?

**Reply by chrjohn** [reply]
There are no special settings, here is a config that worked for us but it is pretty much standard. Also you didn't state what works and what doesn't. Actually log4j2 should log in its default config without specifying anything special. So if that does not work then something is probably broken in your config or classpath. 

```
<?xml version="1.0" encoding="UTF-8"?>
<Configuration>
    <Appenders>
        <Console name="STDOUT" target="SYSTEM_OUT">
            <PatternLayout pattern="[%d] [%-5p] [%c] [%t] %m%n"/>
        </Console>
        <RollingFile name="RollingFile" fileName="${sys:user.home}/log/xxx.log"
                     filePattern="${sys:user.home}/log/oldlog/xxx-%d{yyyyMMdd}-%i.log.gz">
            <PatternLayout>
                <Pattern>[%d] [%-5p] [%c] [%t] %m%n</Pattern>
            </PatternLayout>
            <Policies>
                <TimeBasedTriggeringPolicy/>
                <SizeBasedTriggeringPolicy size="100 MB"/>
            </Policies>
            <DefaultRolloverStrategy>
                <Delete basePath="${sys:user.home}/log/oldlog/" maxDepth="1">
                    <IfFileName glob="*.log.gz" />
                    <IfLastModified age="30d" />
                </Delete>
                <Delete basePath="${sys:user.home}/log/" maxDepth="1">
                    <IfFileName glob="xxx_javaProcess.log">
                        <IfAccumulatedFileSize exceeds="50 MB" />
                    </IfFileName>
                </Delete>
            </DefaultRolloverStrategy>
        </RollingFile>
    </Appenders>

    <Loggers>
        <Root level="debug">
            <AppenderRef ref="STDOUT"/>
        </Root>

        <Logger name="quickfix.mina" level="info" additivity="false">
            <AppenderRef ref="RollingFile"/>
        </Logger>
        <Logger name="org.apache.mina" level="info" additivity="false">
            <AppenderRef ref="RollingFile"/>
        </Logger>
    </Loggers>
</Configuration>
```

**Reply by hamilton-capital** [reply]
Here is a very simple QFJ/log4j2 maven project if you have a few minutes to look at - maybe you can more easily see what the issue is. All I am trying to do here is create QFJ logs (not rolling yet) using log4j2. The logs (quickfixj.event, quickfixj.msg.incoming and quickfixj.msg.outgoing) are never created for some reason - I cant understand why. 

[QFJTest.zip](https://github.com/quickfix-j/quickfixj/files/9933953/QFJTest.zip)

**Reply by chrjohn** [reply]
But your connector starts up successfully? Is the destination writable and does exist? Did you try specifying an absolute path?

**Reply by hamilton-capital** [reply]
The connector starts up perfectly. The destination is writable - I can see a few FIX files that do get created (.body, .header, .senderseqnums, .session and .targetseqnums). 

I don't see the quickfixj.event, quickfixj.msg.incoming and quickfixj.msg.outgoing files created.

**Reply by chrjohn** [reply]
Do you see output on the console?

**Reply by hamilton-capital** [reply]
yes here it is:

Nov 06, 2022 9:42:24 AM quickfix.DefaultSessionSchedule <init>
INFO: [FIX.4.3:ServerUI->ClientUI] weekly, MON 00:00:00-UTC - SUN 23:59:59-UTC
Nov 06, 2022 9:42:24 AM org.slf4j.impl.JDK14LoggerAdapter fillCallerData
INFO: FIX.4.3:ServerUI->ClientUI: Session FIX.4.3:ServerUI->ClientUI schedule is weekly, MON 00:00:00-UTC - SUN 23:59:59-UTC
Nov 06, 2022 9:42:24 AM org.slf4j.impl.JDK14LoggerAdapter fillCallerData
INFO: FIX.4.3:ServerUI->ClientUI: Created session: FIX.4.3:ServerUI->ClientUI
Nov 06, 2022 9:42:24 AM quickfix.mina.SessionConnector startSessionTimer
INFO: SessionTimer started
Nov 06, 2022 9:42:24 AM quickfix.mina.NetworkingOptions logOption
INFO: Socket option: SocketTcpNoDelay=true
Nov 06, 2022 9:42:24 AM quickfix.mina.NetworkingOptions logOption
INFO: Socket option: SocketSynchronousWrites=false
Nov 06, 2022 9:42:24 AM quickfix.mina.NetworkingOptions logOption
INFO: Socket option: SocketSynchronousWriteTimeout=30000
Nov 06, 2022 9:42:24 AM quickfix.mina.acceptor.AbstractSocketAcceptor startAcceptingConnections
INFO: Listening for connections at 0.0.0.0/0.0.0.0:5001 for session(s) [FIX.4.3:ServerUI->ClientUI]
Nov 06, 2022 9:42:24 AM quickfix.mina.SingleThreadedEventHandlingStrategy lambda$blockInThread$1
INFO: Started QFJ Message Processor

**Reply by chrjohn** [reply]
Hmm you only have `Console` appenders in your logging config. Might that be the issue? Output on the console seems to work.

**Reply by hamilton-capital** [reply]
Actually no, when i went back and had a look, the Console Appender in my flog4j2.xml specifies a certain pattern layout. The messages above are ignoring that layout which implies log4j2 is not configured in QFJ. Funnily enough, when I attempt to write a message in my main method (outside of QFJ), that gets printed in the correct format. So its back to QFJ not picking up log4j2. 

The sample project I attached is a stand-alone project. Is it possible you save it locally and try running it to see if we can spot the problem?

**Reply by chrjohn** [reply]
Please paste the output of the contents of your class path of your app.

**Reply by chrjohn** [reply]
As expected, I think you have a problem within your classpath, I got it working with the config I posted earlier and the following additional libs on the classpath (additional to QFJ and the libs in QFJ's lib folder):

```
log4j-slf4j-impl-2.19.0.jar
log4j-api-2.19.0.jar
log4j-core-2.19.0.jar
```


If you still have problems try passing the following java property `-Dlog4j2.debug=true` which will show from where the config is loaded. In my case e.g.
```
DEBUG StatusLogger Configuration XmlConfiguration[location=jar:file:/xxx/QFJTest/target/QFJTest-1.0-SNAPSHOT.jar!/log4j2.xml] initialized
```

The log then contains:

```
cat ~/log/xxx.log 
[2022-11-08 13:09:30,733] [INFO ] [quickfixj.event] [main] FIX.4.3:ServerUI->ClientUI: Session FIX.4.3:ServerUI->ClientUI schedule is daily, 00:00:00-UTC - 00:00:00-UTC
[2022-11-08 13:09:30,735] [INFO ] [quickfixj.event] [main] FIX.4.3:ServerUI->ClientUI: Created session: FIX.4.3:ServerUI->ClientUI
[2022-11-08 13:09:30,736] [INFO ] [quickfix.SocketAcceptor] [main] SessionTimer started
[2022-11-08 13:09:30,755] [INFO ] [quickfix.mina.NetworkingOptions] [main] Socket option: SocketTcpNoDelay=true
[2022-11-08 13:09:30,756] [INFO ] [quickfix.mina.NetworkingOptions] [main] Socket option: SocketSynchronousWrites=false
[2022-11-08 13:09:30,756] [INFO ] [quickfix.mina.NetworkingOptions] [main] Socket option: SocketSynchronousWriteTimeout=30000
[2022-11-08 13:09:30,772] [INFO ] [quickfix.SocketAcceptor] [main] Listening for connections at 0.0.0.0/0.0.0.0:5001 for session(s) [FIX.4.3:ServerUI->ClientUI]
[2022-11-08 13:09:30,774] [INFO ] [quickfix.SocketAcceptor] [QFJ Message Processor] Started QFJ Message Processor
```

**Comment by hamilton-capital** [comment]
Apologies for the delay. 

As far as I can see i am using the same libs as you but now I am getting the following error: 

**SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.**

class path libs.

C:\Users\xxx\.jdks\openjdk-16.0.2\bin\java.exe "-javaagent:
C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2021.1.3\lib\idea_rt.jar=57291:
C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2021.1.3\bin" -Dfile.encoding=UTF-8 -classpath 
C:\Users\xxx\projects\QFJTest\target\classes;
C:\Users\xxx\.m2\repository\org\quickfixj\quickfixj-core\2.3.1\quickfixj-core-2.3.1.jar;
C:\Users\xxx\.m2\repository\org\apache\mina\mina-core\2.1.4\mina-core-2.1.4.jar;
C:\Users\xxx\.m2\repository\org\slf4j\slf4j-api\1.7.30\slf4j-api-1.7.30.jar;
C:\Users\xxx\.m2\repository\org\quickfixj\quickfixj-all\2.3.1\quickfixj-all-2.3.1.jar;
C:\Users\xxx\.m2\repository\org\quickfixj\quickfixj-messages-all\2.3.1\quickfixj-messages-all-2.3.1.jar;
C:\Users\xxx\.m2\repository\org\quickfixj\quickfixj-codegenerator\2.3.1\quickfixj-codegenerator-2.3.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-plugin-api\3.8.1\maven-plugin-api-3.8.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-model\3.8.1\maven-model-3.8.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-artifact\3.8.1\maven-artifact-3.8.1.jar;
C:\Users\xxx\.m2\repository\org\apache\commons\commons-lang3\3.8.1\commons-lang3-3.8.1.jar;
C:\Users\xxx\.m2\repository\org\eclipse\sisu\org.eclipse.sisu.plexus\0.3.4\org.eclipse.sisu.plexus-0.3.4.jar;
C:\Users\xxx\.m2\repository\javax\enterprise\cdi-api\1.0\cdi-api-1.0.jar;
C:\Users\xxx\.m2\repository\javax\annotation\jsr250-api\1.0\jsr250-api-1.0.jar;
C:\Users\xxx\.m2\repository\javax\inject\javax.inject\1\javax.inject-1.jar;
C:\Users\xxx\.m2\repository\org\eclipse\sisu\org.eclipse.sisu.inject\0.3.4\org.eclipse.sisu.inject-0.3.4.jar;
C:\Users\xxx\.m2\repository\org\codehaus\plexus\plexus-component-annotations\1.5.5\plexus-component-annotations-1.5.5.jar;
C:\Users\xxx\.m2\repository\org\codehaus\plexus\plexus-utils\3.2.1\plexus-utils-3.2.1.jar;
C:\Users\xxx\.m2\repository\org\codehaus\plexus\plexus-classworlds\2.6.0\plexus-classworlds-2.6.0.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-project\2.2.1\maven-project-2.2.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-settings\2.2.1\maven-settings-2.2.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-profile\2.2.1\maven-profile-2.2.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-artifact-manager\2.2.1\maven-artifact-manager-2.2.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-repository-metadata\2.2.1\maven-repository-metadata-2.2.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\wagon\wagon-provider-api\1.0-beta-6\wagon-provider-api-1.0-beta-6.jar;
C:\Users\xxx\.m2\repository\backport-util-concurrent\backport-util-concurrent\3.1\backport-util-concurrent-3.1.jar;
C:\Users\xxx\.m2\repository\org\apache\maven\maven-plugin-registry\2.2.1\maven-plugin-registry-2.2.1.jar;
C:\Users\xxx\.m2\repository\org\codehaus\plexus\plexus-interpolation\1.11\plexus-interpolation-1.11.jar;
C:\Users\xxx\.m2\repository\org\codehaus\plexus\plexus-container-default\1.0-alpha-9-stable-1\plexus-container-default-1.0-alpha-9-stable-1.jar;
C:\Users\xxx\.m2\repository\junit\junit\3.8.1\junit-3.8.1.jar;
C:\Users\xxx\.m2\repository\classworlds\classworlds\1.1-alpha-2\classworlds-1.1-alpha-2.jar;
C:\Users\xxx\.m2\repository\net\sf\saxon\Saxon-HE\9.8.0-4\Saxon-HE-9.8.0-4.jar;
C:\Users\xxx\.m2\repository\org\quickfixj\quickfixj-dictgenerator\2.3.1\quickfixj-dictgenerator-2.3.1.jar;
C:\Users\xxx\.m2\repository\jaxen\jaxen\1.2.0\jaxen-1.2.0.jar;
C:\Users\xxx\.m2\repository\org\apache\logging\log4j\log4j-core\2.19.0\log4j-core-2.19.0.jar;
C:\Users\xxx\.m2\repository\org\apache\logging\log4j\log4j-api\2.19.0\log4j-api-2.19.0.jar;
C:\Users\xxx\.m2\repository\org\slf4j\slf4j-reload4j\2.0.3\slf4j-reload4j-2.0.3.jar;
C:\Users\xxx\.m2\repository\ch\qos\reload4j\reload4j\1.2.22\reload4j-1.2.22.jar main -Dlog4j2.debug=true

**Reply by chrjohn** [reply]
Maybe try to add `log4j-slf4j-impl-2.19.0.jar` which is missing from your classpath but mine has it. Also I don't think you need reload4j when you are using log4j2.
I think that is all the help I can offer. Good luck :)

**Reply by hamilton-capital** [reply]
Thank you! It all works now after making those changes. 

One last question, I am noticing a FIX.4.3-xxx.**body** file that gets created when the session starts (also a .header, .senderseqnums, .session and .targetseqnums files). Is there a way of disabling the .body file as I already get the FIX messages in the log4j2 logfile.

**Reply by chrjohn** [reply]
Great, then please mark this discussion as "answered".

The .body file (and the other files you mentioned) cannot be disabled since that is QFJ's internal session store, a `FileStore` in your case. Maybe it is better to have the file store in a directory separate from your log file. The log file is meant to be human readable, the session store mostly is not.
